exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp

```diff
@@ -0,0 +1,219 @@
+    k_gains_.assign(T, Eigen::MatrixXd(u.rows(), x.rows()));
+    ku_gains_.assign(T, Eigen::MatrixXd(u.rows(), u.rows()));
+    kv_gains_.assign(T, Eigen::MatrixXd(u.rows(), x.rows()));
+    vk_gains_.assign(T, Eigen::MatrixXd(x.rows(), x.cols()));
+
+    HIGHLIGHT_NAMED("ILQRSolver", "initialized");
+}
+
+void ILQRSolver::BackwardPass()
+{
+    const double k_min_clamp_ = -1e10;
+    const double k_max_clamp_ = 1e10;
+
+    Eigen::MatrixXd Qf = prob_->get_Qf(), R = prob_->get_R();
+    Eigen::MatrixXd X_star = prob_->get_X_star();
+    DynamicsSolverPtr dynamics_solver = prob_->GetScene()->GetDynamicsSolver();
+
+    // eq. 18
+    vk_gains_[T - 1] = Qf * dynamics_solver->StateDelta(prob_->get_X(T - 1), X_star.col(T - 1));
+    vk_gains_[T - 1] = vk_gains_[T - 1].unaryExpr([k_min_clamp_, k_max_clamp_](double x) -> double {
+        return std::min(std::max(x, k_min_clamp_), k_max_clamp_);
+    });
+
+    Eigen::MatrixXd Sk = Qf;
+    Sk = Sk.unaryExpr([k_min_clamp_, k_max_clamp_](double x) -> double {
+        return std::min(std::max(x, k_min_clamp_), k_max_clamp_);
+    });
+
+    for (int i = T - 2; i > 0; i--)
+    {
+        Eigen::MatrixXd x = prob_->get_X(i), u = prob_->get_U(i), Q = prob_->get_Q(i);
+        Eigen::MatrixXd Ak = dynamics_solver->fx(x, u), Bk = dynamics_solver->fu(x, u);
+
+        // eq. 13-16
+        Ak = Ak * dynamics_solver->get_dt() + Eigen::MatrixXd::Identity(Ak.rows(), Ak.cols());
+        Bk = Bk * dynamics_solver->get_dt();
+        // this inverse is common for all factors
+        Eigen::MatrixXd _inv = (Eigen::MatrixXd::Identity(R.rows(), R.cols()) * 1e-5 + R + Bk.transpose() * Sk * Bk).inverse();
+
+        kv_gains_[i] = _inv * Bk.transpose();
+        k_gains_[i] = _inv * Bk.transpose() * Sk * Ak;
+        ku_gains_[i] = _inv * R;
+        Sk = Ak.transpose() * Sk * (Ak - Bk * k_gains_[i]) + Q;
+
+        vk_gains_[i] = ((Ak - Bk * k_gains_[i]).transpose() * vk_gains_[i + 1]) -
+                     (k_gains_[i].transpose() * R * u) + (Q * x);
+
+        // fix for large values
+        Sk = Sk.unaryExpr([k_min_clamp_, k_max_clamp_](double x) -> double {
+            return std::min(std::max(x, k_min_clamp_), k_max_clamp_);
+        });
+        vk_gains_[i] = vk_gains_[i].unaryExpr([k_min_clamp_, k_max_clamp_](double x) -> double {
+            return std::min(std::max(x, k_min_clamp_), k_max_clamp_);
+        });
+    }
+}
+
+double ILQRSolver::ForwardPass(double alpha, Eigen::MatrixXdRef ref_trajectory)
+{
+    double cost = 0;
+    DynamicsSolverPtr dynamics_solver = prob_->GetScene()->GetDynamicsSolver();
+    for (int i = 0; i < T - 1; ++i)
+    {
+        Eigen::MatrixXd u = prob_->get_U(i);
+        // eq. 12
+        Eigen::MatrixXd delta_uk = -ku_gains_[i] * u - kv_gains_[i] * vk_gains_[i + 1] - k_gains_[i] * dynamics_solver->StateDelta(prob_->get_X(i), ref_trajectory.col(i));
+
+        u = u + alpha * delta_uk;
+        // clamp controls
+        if (parameters_.ControlLimits.size() > 0)
```

Comment:
Would also need to check that size is the same size as u. And please move ControlLimits to dynamics solver (abstract)